Implement testdriver for Firefox using marionette backend

jgraham opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/9391)
<!-- Reviewable:end -->

jgraham committed:
Refactor testdriver implemenation.

This puts the main logic for handling testdriver callbacks into the
base class where it can be shared by multiple implemenations. It also
moves most of the interaction onto the Protocol object which can be
implemented on a per-executor basis.

jgraham committed:
Implement testdriver for marionette.

This brings the implemenatation in line with the Chrome one i.e. a
single click method is supported.

jgraham committed:
Add a basic smoketest for testdriver's click()

jgraham commented:
r? @andreastt @gsnedders

w3c-bots commented:
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/23887)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/337555049)

gsnedders commented:
Reviewed 7 of 7 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 489 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aINbkE-evjySGcBkp:-L4aINbkE-evjySGcBkq:b79jhqr) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/base.py#L489)):*
> ```Python
> 
>     def __init__(self, logger, protocol, test_window):
>         self.protocol = protocol
> ```

Can we state what the type of `protocol` is? I was expecting it to be a Marionette/Selenium object, but it appears to be an executor?

---

*[tools/wptrunner/wptrunner/executors/base.py, line 501 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aFLlk7oy7jwSOy30W:-L4aFLlk7oy7jwSOy30X:bsdsh3e) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/base.py#L501)):*
> ```Python
>         }
> 
>     def __call__(self, result):
> ```

Can we add some comment saying what this assumes the structure of result is? Because it's not at all immediately obvious, especially with various subscripts of `result` being used all over the place.

It's one thing to have this assuming various things about the type of `result` when it's specific to the Selenium executor, but it seems much more confusing when it's supposedly generic.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 513 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aHFFv921V35MOsoj1:-L4aGucP804Hq3YDsGSn:bxstdfj) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/base.py#L513)):*
> ```Python
> 
>     def process_action(self, result):
>         parent = self.protocol.current_window_handle()
> ```

Can we somewhere document things that the protocol object are expected to implement, in an ABC or something?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391)*
<!-- Sent from Reviewable.io -->

andreastt commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -476,3 +476,83 @@ def is_alive(self):
+    def __call__(self, data):
+        selector = data["selector"]
+        elements = self.protocol.find_elements_by_css_selector(selector)
+        print elements
```
Leftover from debug?

andreastt commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -476,3 +476,83 @@ def is_alive(self):
+        elements = self.protocol.find_elements_by_css_selector(selector)
+        print elements
+        if len(elements) == 0:
+            raise ValueError("Selector matches no elements")
```
Do we think one error type is fine, or is there a requirement for more finely grained types?

andreastt commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -476,3 +476,83 @@ def is_alive(self):
+        if len(elements) == 0:
+            raise ValueError("Selector matches no elements")
+        elif len(elements) > 1:
+            raise ValueError("Selector matches multiple elements")
```
Couldn’t this use `find_element_by_css_selector` above? Or is the concern that it is fallible and might return an error?

andreastt commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -476,3 +476,83 @@ def is_alive(self):
+            self.protocol.click_element(elements[0])
+        except Exception as e:
+            self.logger.warning(e)
+            return False
```
Should the error not be propagated?

andreastt commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -306,13 +333,33 @@ def clear_origin(self, url):
             self.marionette.execute_script(script)
 
+    def find_elements_by_css_selector(self, selector):
+        return self.marionette.find_elements("css selector", selector)
```
Instead of these, couldn’t you just called `self.protocol.marionette.find_elements` so you don’t have to reinvent the API?

andreastt commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -306,13 +333,33 @@ def clear_origin(self, url):
+        }
+        if message:
+            obj["message"] = str(message)
+        self.marionette.execute_script("window.postMessage(%s, '*')" % json.dumps(obj))
```
This is clever, I like it.

andreastt commented on tools/wptrunner/wptrunner/executors/testharness_marionette.js:
```diff
@@ -1,31 +1,24 @@
-    clearTimeout(timer);
-    var tests = event.data.tests;
-    var status = event.data.status;
+window.wrappedJSObject.message_queue = [];
```
Do you still need `wrappedJSObject`?

andreastt commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -384,6 +431,8 @@ def _run(self):
 
 
 class MarionetteTestharnessExecutor(TestharnessExecutor):
+    supports_testdriver = True
```
Where does this get picked up?

andreastt commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -476,3 +476,83 @@ def is_alive(self):
+            raise ValueError("Unknown callback type %r" % result[1])
+
+    def process_complete(self, result):
+        rv = [result[0]] + result[2]
```
I think it’s worth documenting the message format somewhere. Reverse engineering the test data format when I implemented the pytest runner was not my greatest fun-packed day at work.

jgraham commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 489 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aINbkE-evjySGcBkp:-L4c4maNCaXrIBcFcqc9:b-rjez44) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/base.py#L489)):*
<details><summary><i>Previously, gsnedders (Geoffrey Sneddon) wrote…</i></summary><blockquote>

Can we state what the type of `protocol` is? I was expecting it to be a Marionette/Selenium object, but it appears to be an executor?
</blockquote></details>

It's a `Protocol` object. But yes. we can.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 513 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aHFFv921V35MOsoj1:-L4c3wD8qozfn3BPyAXx:b-y35png) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/base.py#L513)):*
<details><summary><i>Previously, gsnedders (Geoffrey Sneddon) wrote…</i></summary><blockquote>

Can we somewhere document things that the protocol object are expected to implement, in an ABC or something?
</blockquote></details>

I think if we want to do this the only sensible option is to pull it apart into multiple objects and compose them together somehow. At the moment a protocol is a grab-bag of things that use whatever the underlying wire protocol is.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 550 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVq-r1-550:-L4c4epS_-pQ_AOu7F_i:bg70cd1) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/base.py#L550)):*
<details><summary><i>Previously, andreastt (Andreas Tolfsen) wrote…</i></summary><blockquote>

Couldn’t this use `find_element_by_css_selector` above? Or is the concern that it is fallible and might return an error?
</blockquote></details>

YEah, it's that >1 shoudl return an error.

---

*[tools/wptrunner/wptrunner/executors/executormarionette.py, line 337 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVr-r1-337:-L4c4FHuEPqT6P631lMq:b367mpq) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/executormarionette.py#L337)):*
<details><summary><i>Previously, andreastt (Andreas Tolfsen) wrote…</i></summary><blockquote>

Instead of these, couldn’t you just called `self.protocol.marionette.find_elements` so you don’t have to reinvent the API?
</blockquote></details>

I think we decided this wasn't an issue

---

*[tools/wptrunner/wptrunner/executors/testharness_marionette.js, line 4 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVt-r1-4:-L4c4TXrLma6xOiey8gn:bcnx4h) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/testharness_marionette.js#L4)):*
<details><summary><i>Previously, andreastt (Andreas Tolfsen) wrote…</i></summary><blockquote>

Do you still need `wrappedJSObject`?
</blockquote></details>

I don't know! If not then we can share scripts with the selenium implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391)*
<!-- Sent from Reviewable.io -->

andreastt commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/executors/testharness_marionette.js, line 4 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVt-r1-4:-L4erDMucKislkfhSVkq:b-qr5sy7) ([raw file](https://github.com/w3c/web-platform-tests/blob/af20fe0a467348acb8ee3c0474704f0aa7a085d0/tools/wptrunner/wptrunner/executors/testharness_marionette.js#L4)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

I don't know! If not then we can share scripts with the selenium implementation.
</blockquote></details>

I suspect you could do that. But I’m fine with making that change at a later point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391)*
<!-- Sent from Reviewable.io -->

jgraham committed:
Refactor the Protocol object into multiple ProtocolParts.

The goal here is to make it possible for different implemenations to
implement defined subsets of the Protocol object, but require each
subset to match a specific base class so it's clear when you have
implemented all the pieces of a specific subset.

The code here uses a sort of composition rather than
inheritance. Multiple inheritance tends to be rather confusing, so
instead we do something similar but slihgtly more explicit; assigning
each ProtocolPart to an attibute of the Protocol so that the methods
on a part with name foo are available as protocol.foo.method. This
involves some changes to the lifecycle methods. In particular setup()
is now split into connect() and after_connect() that are expected to
be implemented in base classes. Between these being called we call
setup() on each ProtocolPart so that we may give them direct access to
a connection object or similar.

kereliuk commented:
Review status: 4 of 11 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 489 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aINbkE-evjySGcBkp:-L6RrvEq40HCsKgGViK0:b-fqk7q7) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/base.py#L489)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

It's a `Protocol` object. But yes. we can.
</blockquote></details>

+1 to this. Although I don't think we should be draconian about it, I think we should air on the side of documenting the types of stuff everywhere we can since this is python :)

---

*[tools/wptrunner/wptrunner/executors/base.py, line 509 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVq-r1-509:-L6RsVEN4txbK4qK7wA_:bpmyk56) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/base.py#L509)):*
<details><summary><i>Previously, andreastt (Andreas Tolfsen) wrote…</i></summary><blockquote>

I think it’s worth documenting the message format somewhere. Reverse engineering the test data format when I implemented the pytest runner was not my greatest fun-packed day at work.
</blockquote></details>

I have also had this same problem :)

---

*[tools/wptrunner/wptrunner/executors/base.py, line 513 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aHFFv921V35MOsoj1:-L6RskA81UwNhhPp_uK7:bxym7uz) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/base.py#L513)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

I think if we want to do this the only sensible option is to pull it apart into multiple objects and compose them together somehow. At the moment a protocol is a grab-bag of things that use whatever the underlying wire protocol is.
</blockquote></details>

AFAIK the protocol is expected to implement:
current_window_handle
execute_sync
execute_async
switch_to_window
new_session
get_elements_by_css (for testdriver.js)

but in the future the plan is to have many more things implemented for testdriver, so I am +1 to at least documenting what things are currently expected somewhere

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391)*
<!-- Sent from Reviewable.io -->

kereliuk commented:
Also are you able to write tests for at least part of this (maybe the protocol file?). We currently don't have much coverage test wise with wptrunner, but I want to change that :)

jgraham commented:
Review status: 4 of 11 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 501 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aFLlk7oy7jwSOy30W:-L6RybrSE1_g0HbNIamG:b-hgem35) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/base.py#L501)):*
<details><summary><i>Previously, gsnedders (Geoffrey Sneddon) wrote…</i></summary><blockquote>

Can we add some comment saying what this assumes the structure of result is? Because it's not at all immediately obvious, especially with various subscripts of `result` being used all over the place.

It's one thing to have this assuming various things about the type of `result` when it's specific to the Selenium executor, but it seems much more confusing when it's supposedly generic.
</blockquote></details>

I didn't add a comment, but I made the structure explicit in the usage.

---

*[tools/wptrunner/wptrunner/executors/base.py, line 513 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aHFFv921V35MOsoj1:-L6RzC2Yv_zzuQVCEMLv:bu0facx) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/base.py#L513)):*
<details><summary><i>Previously, kereliuk (Jonathon Kereliuk) wrote…</i></summary><blockquote>

AFAIK the protocol is expected to implement:
current_window_handle
execute_sync
execute_async
switch_to_window
new_session
get_elements_by_css (for testdriver.js)

but in the future the plan is to have many more things implemented for testdriver, so I am +1 to at least documenting what things are currently expected somewhere
</blockquote></details>

I'm not really sure what the documentation would look like here. What you need to implement depends on what features you support, which in turn depends on your executor.

A long term consequence of the design here is that we could remove some of the duplication between executors and put all the protocol-specific logic in the Protocol object. I haven't done that refactor yet because I would like to land this change at some point ;)

---

*[tools/wptrunner/wptrunner/executors/base.py, line 548 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVq-r1-548:-L4r18a5po8kBbBVoJpp:bybmchs) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/base.py#L548)):*
<details><summary><i>Previously, andreastt (Andreas Tolfsen) wrote…</i></summary><blockquote>

Do we think one error type is fine, or is there a requirement for more finely grained types?
</blockquote></details>

I think that it doesn't matter much because these aren't going to be handled, they're just going to be reported to the user.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/9391)*
<!-- Sent from Reviewable.io -->

jgraham commented:
The protocol file seems hard to test in isolation since it requires an actual browser connection to test.

My preferred way of testing this is to write wptrunner integration tests, check them in to `infrastructure` and then land #9676 so that they are run in Travis (note this PR has such a test for `click`).

gsnedders commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/executors/executormarionette.py, line 434 at r1](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-L4aD2Dz7ftX4taeauVr-r1-434:-L7eAnYgCnFtRKMVY85y:b-q9krs3) ([raw file](https://github.com/w3c/web-platform-tests/blob/774770362e9d1c1cc1c687e1407f7666e2fcf12d/tools/wptrunner/wptrunner/executors/executormarionette.py#L434)):*
<details><summary><i>Previously, andreastt (Andreas Tolfsen) wrote…</i></summary><blockquote>

Where does this get picked up?
</blockquote></details>

`tools/wptrunner/wptrunner/wptrunner.py`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/9391#-:-L7eBAX36U9VX21CCAOZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->
